sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SchemaPruning.scala

```diff
@@ -204,5 +205,58 @@ object SchemaPruning extends Rule[LogicalPlan] {
     }
   }
 
+  private def applyMetadataSchemaPruning(plan: LogicalPlan): LogicalPlan =
+    plan transformDown {
+      case op @ PhysicalOperation(projects, filters, l @ LogicalRelation(_, _, _, _))
+        if containsMetadataAttributes(l) => pruneMetadataSchema(l, projects, filters).getOrElse(op)
+    }
 
+  /**
+   * This method returns optional logical plan with pruned metadata schema.
+   * `None` is returned if no nested field is required or all nested fields are required.
+   */
+  private def pruneMetadataSchema(
```

Comment:
The code structure is very similar to `prunePhysicalColumns`. Can we combine these two, and prune data and metadata schema in one pass?

Reply:
I can definitely try but not sure if it will cause more complexity: like no need to do `leafNodeBuilder` for metadata, dataSchema pruning is only for Parquet + ORC and controlled by a flag, etc.

Reply:
> like no need to do leafNodeBuilder

It's needed now, as you did in L248: `relation.copy(output = prunedOutput)`

Reply:
ah ok, let me try to merge those two.